### PR TITLE
test: USDT tracepoint interface tests

### LINKED ISSUE #23296: tracing: tests for USDT tracepoints
#22006 added the first three USDT based tracepoints to Bitcoin Core. To provide a [semi-stable](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#semi-stable-api) tracepoint API the tracepoints need test coverage. The tracepoints can be tested in the functional tests using the Python wrapper of [BCC](https://github.com/iovisor/bcc). 

Before adding more tracepoints, the existing three tracepoints from #22006 should be tested.

Notes:
1. We currently only support the tracepoints on Linux. The tests should be skipped on other operating systems.
2. Hooking into the tracepoints via the Linux kernel requires special privileges. Since kernel version 5.8. (Aug. 2020) the `CAP_BPF` can be used. On older kernel version the overloaded catch-all capability `CAP_SYS_ADMIN` is required. Functional tests shouldn't require `CAP_SYS_ADMIN` as that essentially means running the test suite with `root` privileges.  
3. The tests require the BCC Python library. This should be an optional dependency. Tests should be skipped if the dependency isn't present. 

The `connect_block` tracepoint can be tested by mining blocks with transactions and checking that the tracepoint passes the correct data. The `net` `inbound_message` and `outbound_message` tracepoints can be tested by checking the traffic between two nodes.

### PR DESCRIPTION
**Pull request by 0xB10C**
This adds functional tests for the USDT tracepoints added in https://github.com/bitcoin/bitcoin/pull/22006 and https://github.com/bitcoin/bitcoin/pull/22902. This partially fixes #23296. The tests **are probably skipped** on most systems as these tests require:
- a Linux system with a kernel that supports BPF (and available kernel headers)
- that Bitcoin Core is compiled with tracepoints for USDT support (default when compiled with depends)
- [bcc](https://github.com/iovisor/bcc) installed
- the tests are run with a privileged user that is able to e.g. do BPF syscalls and load BPF maps

The tests are not yet run in our CI as the CirrusCI containers lack the required permissions (see https://github.com/bitcoin/bitcoin/issues/23296#issuecomment-1024920845). Running the tests in a VM in the CI could work, but I haven't experimented with this yet. The priority was to get the actual tests done first to ensure the tracepoints work as intended for the v23.0 release. Running the tracepoint tests in the CI is planned as the next step to finish #23296.

The tests can, however, be run against e.g. release candidates by hand. Additionally, they provide a starting point for tests for future tracepoints. PRs adding new tracepoint should include tests. This makes reviewing these PRs easier.

The tests require privileges to execute BPF sycalls (`CAP_SYS_ADMIN` before Linux kernel 5.8 and `CAP_BPF` and `CAP_PERFMON` on 5.8+) and permissions to `/sys/kernel/debug/tracing/`. It's currently recommended to run the tests in a virtual machine (or on a VPS) where it's sensible to use the `root` user to gain these privileges. Never run python scripts you haven't carefully reviewed with `root` permissions! It's unclear if a non-root user can even gain the required privileges. This needs more experimenting.

The goal here is to test the tracepoint interface to make sure the [documented interface](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#tracepoint-documentation) does not break by accident. The tracepoints expose implementation details. This means we also need to rely on implementation details of Bitcoin Core in these functional tests to trigger the tracepoints. An example is the test of the `utxocache:flush` tracepoint: On Bitcoin Core shutdown, the UTXO cache is flushed twice. The corresponding tracepoint test expects two flushes, too - if not, the test fails. Changing implementation details could cause these tests to fail and the tracepoint API to break. However, we purposefully treat the tracepoints only as [**semi-stable**](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#semi-stable-api). The tracepoints should not block refactors or changes to other internals.

### CODE REVIEW

**Comment by laanwj** [comment]
Nice! Concept ACK.

**Comment by 0xB10C** [comment]
#23907 is merged. Rebased and ready for review.

**Comment by theStack** [comment]
Concept ACK

**Comment by jb55** [comment]
Tests pass on my end, but oddly enough when I run `bpftrace contrib/tracing/log_utxos.bt` I get a segfault. anyone else seeing this?

```
Thread 1 "bpftrace" received signal SIGSEGV, Segmentation fault.
0x00007fffefc816e4 in llvm::PointerType::get(llvm::Type*, unsigned int) () from /nix/store/sfydnaab0wxn2qm3pkaab5x1fcagzxpf-llvm-11.1.0-lib/lib/libLLVM-11.so
(gdb) bt
#0  0x00007fffefc816e4 in llvm::PointerType::get(llvm::Type*, unsigned int) ()
   from /nix/store/sfydnaab0wxn2qm3pkaab5x1fcagzxpf-llvm-11.1.0-lib/lib/libLLVM-11.so
#1  0x000000000055fc37 in bpftrace::ast::CodegenLLVM::createFormatStringCall(bpftrace::ast::Call&, int&, std::vector<std::tuple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<bpftrace::Field, std::allocator<bpftrace::Field> > >, std::allocator<std::tuple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<bpftrace::Field, std::allocator<bpftrace::Field> > > > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bpftrace::AsyncAction) ()
#2  0x0000000000569c87 in bpftrace::ast::CodegenLLVM::visit(bpftrace::ast::Call&) ()
#3  0x0000000000558dd6 in bpftrace::ast::CodegenLLVM::accept(bpftrace::ast::Node*) ()
#4  0x0000000000558ee7 in bpftrace::ast::CodegenLLVM::visit(bpftrace::ast::ExprStatement&) ()
#5  0x0000000000558dd6 in bpftrace::ast::CodegenLLVM::accept(bpftrace::ast::Node*) ()
#6  0x000000000056edce in bpftrace::ast::CodegenLLVM::generateProbe(bpftrace::ast::Probe&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, llvm::FunctionType*, bool, std::optional<int>, bool) ()
#7  0x000000000056fbaf in bpftrace::ast::CodegenLLVM::visit(bpftrace::ast::Probe&) ()
#8  0x0000000000558dd6 in bpftrace::ast::CodegenLLVM::accept(bpftrace::ast::Node*) ()
#9  0x0000000000558f6e in bpftrace::ast::CodegenLLVM::visit(bpftrace::ast::Program&) ()
#10 0x0000000000558dd6 in bpftrace::ast::CodegenLLVM::accept(bpftrace::ast::Node*) ()
#11 0x0000000000558feb in bpftrace::ast::CodegenLLVM::generate_ir() ()
#12 0x000000000043d622 in main ()
```

**Comment by 0xB10C** [comment]
@jb55 which version of `bpftrace` are you running? I've got the segfault on 0.13 and 0.14 too, it works fine with 0.12. The issue seems to be related to the printf in the unroll followed by another printf. 

This is kinda off topic here, but there have been quite a few problems with `bpftrace` on different versions. People reported on IRC that the example scripts don't work with their `bpftrace` version. Have been thinking about dropping the examples until `bpftrace` is more stable.

**Comment by jb55** [comment]
tACK 76c60d7b31ccc50b226cdbc5e38be0bd67603408

**Comment by laanwj** [comment]
> It's unclear if a non-root user can even gain the required privileges. This needs more experimenting.

I managed to get the required privileges as a user, and get the tests to run:
```
# chmod 755 /sys/kernel/debug # might want to assign a special group or such
# chmod 755 /sys/kernel/debug/tracing
# chmod 666 /sys/kernel/debug/tracing/uprobe_events
# echo "3" > /proc/sys/kernel/perf_event_paranoid # 3 or lower will do, 4 won't
# setpriv --ambient-caps +cap_38,+cap_39 --inh-caps +cap_38,+cap_39 --init-groups --reuid=1000 --regid=1000 bash
$ getpcaps $$
310068: cap_perfmon,cap_bpf=eip
$ cd …/bitcoin
$ test/functional/interface_usdt_net.py
(passes)
$ test/functional/interface_usdt_validation.py
(passes)
$ test/functional/interface_usdt_utxocache.py
(fails on some probably unrelated assertion)
```

It's apparently also possible to assign fixed capabilities to users through `/etc/security/capability.conf`in some environments. I haven't tried this yet.

BTW: `test_runner.py` fails on all the new tests for me. Although the test itself passes, warnings are printed to stderr, I think causing it to be marked as failed:
```
stderr:
In file included from <built-in>:2:                                                                       
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:80:
include/linux/compiler-clang.h:41:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefine
d]
#define __HAVE_BUILTIN_BSWAP32__
        ^
<command line>:4:9: note: previous definition is here 
#define __HAVE_BUILTIN_BSWAP32__ 1
        ^
…
```

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -821,6 +822,29 @@ def skip_if_no_py3_zmq(self):
...
+    def skip_if_no_bitcoind_tracepoints(self):
+        """Skip the running test if bitcoind has not been compiled with USDT tracepoint support."""
+        if not self.is_usdt_compiled():
+            raise SkipTest("bitcoind has not been built with USDT tracepoints enabled.")
+
+    def skip_if_no_bpf_permissions(self):
+        """Skip the running test if we don't have permissions to do BPF syscalls and load BPF maps."""
+        # check for 'root' permissions
+        if os.geteuid() != 0:
+            raise SkipTest("no permissions to use BPF (please review the tests carefully before running them with higher privileges)")
```
Comment (laanwj) [inline test/functional/test_framework/test_framework.py:841]: As alternative to / in addition to `os.geteuid() != 0` for non-root use we'll probably want to check `CapEff` in `/proc/<pid>/status` for `(1 << CAP_PERFMON) | (1 << CAP_BPF)`.
```
CAP_PERFMON = 38
CAP_BPF = 39
```
Not necessarily in this PR though.

**Comment by jamesob** [comment]
Concept ACK. Nice functional demonstration of USDT usage, too. I'll test this soon.

**Comment by laanwj** [comment]
> Concept ACK. Nice functional demonstration of USDT usage, too. I'll test this soon.

@jamesob I'm going ahead and merge this as I think it's ready. I hope you'll still get around to testing it though.

**Pull request merged**